Rx/v2/src/rxcpp/operators/rx-repeat.hpp

```diff
@@ -42,87 +42,144 @@ struct repeat_invalid : public rxo::operator_base<repeat_invalid_arguments<AN...
-    struct values
-    {
-        values(source_type s, count_type t)
-            : source(std::move(s))
-            , remaining(std::move(t))
-            , repeat_infinitely(t == 0)
-        {
-        }
-        source_type source;
-        count_type remaining;
-        bool repeat_infinitely;
-    };
-    values initial;
 
-    repeat(source_type s, count_type t)
-        : initial(std::move(s), std::move(t))
-    {
+    struct values {
+      values(source_type s, count_type t)
+        : source(std::move(s)),
+          remaining_(std::move(t)) {
+      }
+
+      inline bool completed_predicate() const {
+        // Return true if we are completed
+        return remaining_ <= 0;
+      }
+      
+      inline void on_completed() {
+        // Decrement counter
+        --remaining_;
+      }
+
+      source_type source;
+
+    private:
+      // Counter to hold number of times remaining to complete
+      count_type remaining_;
+    };
+    
+    finite(source_type s, count_type t)
+      : initial_(std::move(s), std::move(t)) {
     }
 
     template<class Subscriber>
     void on_subscribe(const Subscriber& s) const {
-
-        typedef Subscriber output_type;
-        struct state_type
-            : public std::enable_shared_from_this<state_type>
-            , public values
-        {
-            state_type(const values& i, const output_type& oarg)
-                : values(i)
-                , source_lifetime(composite_subscription::empty())
-                , out(oarg)
-            {
-            }
-            composite_subscription source_lifetime;
-            output_type out;
-            composite_subscription::weak_subscription lifetime_token;
-
-            void do_subscribe() {
-                auto state = this->shared_from_this();
-                
-                state->out.remove(state->lifetime_token);
-                state->source_lifetime.unsubscribe();
-
-                state->source_lifetime = composite_subscription();
-                state->lifetime_token = state->out.add(state->source_lifetime);
-
-                state->source.subscribe(
-                    state->out,
-                    state->source_lifetime,
-                // on_next
-                    [state](T t) {
-                        state->out.on_next(t);
-                    },
-                // on_error
-                    [state](std::exception_ptr e) {
-                        state->out.on_error(e);
-                    },
-                // on_completed
-                    [state]() {
-                        if (state->repeat_infinitely || (--state->remaining > 0)) {
-                            state->do_subscribe();
-                        } else {
-                            state->out.on_completed();
-                        }
-                    }
-                );
-            }
-        };
-
+      typedef state_type<values, Subscriber, T> state_t;
+      if (!initial_.completed_predicate()) {
```

Comment:
I thought about it a lot, the question is when should we call on_completed()? There were different versions, return on_completed() immediately. Or for example 0f0f23893a048cb7f104e31b069df6b55d4a1087 used to put on_completed() into sequence but only when the input sequence is fully consumed. So I left it for a future consideration. If you have a strong opinion when on_completed() should be called immedieately, let's put it there.